text/0000-generic-integers.md

```diff
@@ -0,0 +1,437 @@
+
+* Users can always *increase* the alignment and size of integers by wrapping them in `repr(align(...))` structs, but they cannot *decrease* them. This is an argument for making the size/alignment of these integers as small as possible.
+* Loading unaligned integers (lower than the alignment of `usize`) can be particularly slow and require additional operations. However, once these integers are loaded, the backend doesn't need to do any extra operations until they need to be stored again. This is an argument for making the size/alignment of these integers larger than necessary.
+* Right now, C's [`_BitInt`] (on x86_64 Unix) does prefer the second option, but the actual status of the spec is in flux due to [incompatibilities with the ABI for `__int128`][`__int128`], and it's unclear what the final version will be. Whether we should always match the ABI for C depends on what the community prefers most, and whether wrapping values in a struct at FFI boundaries is reasonable.
+
+[`_BitInt`]: https://en.cppreference.com/w/c/language/arithmetic_types
+[`__int128`]: https://gitlab.com/x86-psABIs/x86-64-ABI/-/issues/11
+
+## Limits on `N`
+
+There are two primary limits that restrict how large `N` can be:
+
+1. All allocations in Rust are limited to `isize::MAX` bytes.
+2. Most integer methods and constants use `u32` when counting bits.
+
+The first restriction doesn't matter since `isize::MAX` bytes is `isize::MAX * 8` bits, which is larger than `usize::MAX = isize::MAX * 2 + 1` bits.
+
+However, the second restriction is somewhat significant: for systems where `usize::MAX > u32::MAX`, we are still effectively restricted to `N <= u32::MAX` unless we wish to change these APIs. We can treat this as effectively a post-monomorphisation error similar to the error you might get when adding very large arrays inside your type; it's unlikely that someone might encounter them, but they do exist and have to be accounted for.
+
+It's worth noting that `u32::MAX` bits is equivalent to 0.5 GiB, and thus no integer in Rust will be able to be larger than this amount. This is seen as acceptable because at that size, people can just use their own big-integer types. For now, adding a dedicated big-integer type to the standard library is left as a potential future change.
+
+The compiler should be allowed to restrict `N` even further, maybe even as low as `u16::MAX`, due to other restrictions that may apply. For example, the LLVM backend currently only allows integers with widths up to `u<23>::MAX` (not a typo; 23, not 32). On 16-bit targets, using `usize` further restricts these integers to `u16::MAX` bits.
+
+While `N` could be a `u32` instead of `usize`, keeping it at `usize` makes things slightly more natural when converting bits to array lengths and other length-generics, and these quite high cutoff points are seen as acceptable. In particular, this helps using `N` for an array index until [`generic_const_exprs`] is stabilized.
```

Comment:
Oh, it's absolutely still awkward. I guess that the justification is pretty weak here, although another one is the fact that `usize` means you don't have the extra limit at `isize::MAX` bytes for 16-bit targets.

And as always, you can easily cast the const generic in expressions, but it's harder to do so when you want to reuse it as another generic. But I get the argument that you probably won't want to reuse it anyway, and always will need some sort of computation.

Reply:
I don't think there is any realistic use case for anything close to `u<i16::MAX>`, even on 64-bit targets, so whatever integer type makes typical usage work without casts would be best. While arithmetic operations can be extrapolated to any bit width, once you're past several thousand bits a primitive integer type stops working well. Just use a "proper" bitint library, or write your bit vector ops in terms of `[u64; N]`. It's true that having a post-mono error right above N = 128 is a hassle, but there has to be *some* limit and it should be well below the point where LLVM grinds to a halt because it legalized every bitwise operation into many thousands of word-sized operations and such large basic blocks trigger pathological compile-time performance all over the place.

Reply:
Yeah, I agree with you. It's actually why I mentioned that `u16::MAX` (or indeed, `i16::MAX`) is probably the maximum we should guarantee that's portable, since even at that point you're much better off with a bigint type. 32k is a ridiculously large amount of bits for one number.

Reply:
Script:
```zig
const v: uX = 0;
std.debug.print("{}", .{v});
```
Compile time:
```text
1024 6.129250764846802
2048 6.318152189254761
3072 6.924046754837036
4096 7.588716268539429
5120 10.228423833847046
6144 14.128804206848145
7168 20.459102153778076
8192 29.913185834884644
```
For reference, in a debug build of Zig without cache in my `Windows11`, the compile time increases drastically when bit width is larger than `4096`.
So I think the maximum could be even much smller for better experiement under current LLVM.

Reply:
imo 512 should be more than enough for a MVP

Reply:
Zig has 65535 as the maximum bit width, clang 18 on x86_64 has 8388608.